storage/namespace.go

```diff
@@ -373,24 +379,39 @@ func (n *dbNamespace) Repair(repairer databaseShardRepairer) error {
 		throttlePerShard = time.Duration(int64(repairer.Options().RepairThrottle()) / int64(numShards))
 	}
 
+	workers := pool.NewWorkerPool(repairer.Options().RepairShardConcurrency())
+	workers.Init()
 	for _, shard := range shards {
-		metadataRes, err := shard.Repair(n.id, repairer)
-		if err != nil {
-			multiErr = multiErr.Add(err)
-		} else {
-			numShardsRepaired++
-			numTotalSeries += metadataRes.NumSeries
-			numTotalBlocks += metadataRes.NumBlocks
-			numSizeDiffSeries += metadataRes.SizeDifferences.NumSeries()
-			numSizeDiffBlocks += metadataRes.SizeDifferences.NumBlocks()
-			numChecksumDiffSeries += metadataRes.ChecksumDifferences.NumSeries()
-			numChecksumDiffBlocks += metadataRes.ChecksumDifferences.NumBlocks()
-		}
-		if throttlePerShard > 0 {
-			time.Sleep(throttlePerShard)
-		}
+		shard := shard
+
+		wg.Add(1)
+		workers.Go(func() {
+			defer wg.Done()
+
+			metadataRes, err := shard.Repair(n.id, repairer)
+
+			mutex.Lock()
+			if err != nil {
+				multiErr = multiErr.Add(err)
+			} else {
+				numShardsRepaired++
```

Comment:
Also this happens once per shard, there's only ever a few shards being repaired at any one time.  It's super low contention.